File: 0131-palindrome-partitioning/0131-palindrome-partitioning.js
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function(s) {
    let n= s.length
    let ans = [];
    function isPalindrome(str, left, right) {
        while (left < right) {
            if (str[left++] !== str[right--]) return false;
        }
        return true;
    }
     function backTrack(start,subS){
        if(start == n){
            ans.push([...subS]);
            return;
        }
        
        for(let i =start; i < n; i++){
        //check valid or not for next
            if(isPalindrome(s, start, i)){
            subS.push(s.slice(start, i+1));
            backTrack(i+1,subS);
            subS.pop();
            }
        }
     }
     backTrack(0,[]);
     return ans;
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> ans = new ArrayList<>();
        List<String> temp = new ArrayList<>();
        backtrack(s,temp,0,ans);
        return ans;
    }
    public static void backtrack(String s,List<String> temp, int x,List<List<String>>ans ){
        if(x == s.length()){
            ans.add(new ArrayList<>(temp));
            return;
        }
        for(int i = x; i < s.length() ; i++){
            String sub = s.substring(x,i+1);
            if(isPalindrome(sub)){
                temp.add(sub);
                backtrack(s,temp,i+1,ans);
                temp.remove(temp.size()-1);
            }
        }
    }
    public static boolean isPalindrome(String sub){
        int left = 0;
        int right = sub.length() - 1;

        while (left < right) {
            if (sub.charAt(left) != sub.charAt(right)) {
                return false; 
            }
            left++;
            right--;
        }
        return true; 
    }
}